src/util/evalUnits.ts

```diff
@@ -53,19 +63,51 @@ export enum EvalMessages {
  * the first is the actual calculated number (or null if none) while
  * the second is the message that should appear in case of error
  */
-export function evalUnits(input: string): [number | null, string] {
-  if (!floats.test(input) && !ints.test(input) && !alphaInts.test(input) && !alphaFloats.test(input)) {
+export function evalUnits(input: string, chainDecimals: number): [BN | null, string] {
+  //sanitize input to remove + char if exists
+  input = input && input.replace('+', '');
+  if (
+    !new RegExp(floats).test(input) &&
+    !new RegExp(ints).test(input) &&
+    !new RegExp(alphaInts).test(input) &&
+    !new RegExp(alphaFloats).test(input)
+  ) {
     return [null, EvalMessages.GIBBERISH];
   }
-
-  if (floats.test(input) || ints.test(input)) {
-    return [parseFloat(input.replace(/[,]/g, '.')), EvalMessages.SUCCESS];
-  } else if (alphaInts.test(input) || alphaFloats.test(input)) {
-    const numericPart = parseFloat(input.replace(/[,]/g, '.'));
-    const charPart = input.replace(/[0-9.,]/g, '');
-    const siVal = si.find((s) => s.symbol === charPart);
-    return siVal ? [numericPart * siVal.value, EvalMessages.SUCCESS] : [null, EvalMessages.SYMBOL_ERROR];
-  } else {
-    return [null, EvalMessages.GENERAL_ERROR];
+  // find the character from the alphanumerics
+  const symbol = input.replace(/[0-9.,]/g, '');
+  // find the value from the si list
+  const siVal = si.find((s) => s.symbol === symbol);
+  const numberStr = input.replace(symbol, '').replace(',', '.');
+  let numeric: BN = new BN(0);
+  if (siVal) {
+    const containDecimal = input.replace(/[,]/g, '.').includes('.');
+    if (containDecimal) {
+      const [decPart, fracPart] = numberStr.replace(/[,]/g, '.').split('.');
+      const decimals = fracPart.length;
+      const exp = new BN(10).pow(new BN(decimals));
+      if (symbol.toUpperCase() === symbol || symbol === 'k') {
+        numeric = new BN(new BN(decPart).mul(exp).add(new BN(fracPart)))
+          .mul(new BN(10).pow(new BN(chainDecimals)))
+          .mul(siVal.value)
+          .div(exp);
+      } else {
+        numeric = new BN(new BN(decPart).mul(exp).add(new BN(fracPart)))
+          .mul(new BN(10).pow(new BN(chainDecimals)))
+          .div(siVal.value)
+          .div(exp);
+      }
+    } else {
+      if (symbol.toUpperCase() === symbol || symbol === 'k') {
+        numeric = new BN(new BN(numberStr).mul(new BN(10).pow(new BN(chainDecimals)))).mul(siVal.value);
+      } else {
+        numeric = new BN(new BN(numberStr).mul(new BN(10).pow(new BN(chainDecimals)))).div(siVal.value);
+      }
+    }
+    if (numeric.eq(new BN(0))) {
+      return [null, EvalMessages.ZERO];
+    }
+    return [numeric, EvalMessages.SUCCESS];
   }
+  return [null, EvalMessages.SYMBOL_ERROR];
```

Comment:
I find this much easier to follow and it avoids a lot of repetition.
 
```suggestion
  if (!siVal) {
    return [null, EvalMessages.SYMBOL_ERROR];
  }
  const decimalsBn = new BN(10).pow(new BN(chainDecimals));
  const containDecimal = numberStr.includes('.');
  if (containDecimal) {
      const [decPart, fracPart] = numberStr.split('.');
      const fracDecimals = fracPart.length;
      const fracExp = new BN(10).pow(new BN(fracDecimals));
      numeric = new BN(decPart).mul(exp).add(new BN(fracPart)).mul(decimalsBn).div(fracExp);
    } else {
      numeric = new BN(numberStr).mul(decimalsBn);
    }
    
    // [ToDr] I honestly hate that logic of detecting the symbol by lower/uppercase, but can live with it and won't complain any more ;)
    const isMulSymbol = symbol.toUpperCase() === symbol || symbol === 'k';
    if (isMulSymbol) {
        numeric = numeric.mul(siVal.value);
    } else {
        numeric = numeric.div(siVal.value);
    }
    
    if (numeric.eq(new BN(0))) {
      return [null, EvalMessages.ZERO];
    }
    
    return [numeric, EvalMessages.SUCCESS];
  }

```

Reply:
This code will not work. In addition I think it is better to compare on an existing entity than introducing a new one just to compare on it - but I'll proceed with that even though I do not agree (and you should always complain about what you believe is correct)

Reply:
> This code will not work

Care to elaborate?

> In addition I think it is better to compare on an existing entity than introducing a new one just to compare on it

Could you give an example? I don't think I get it.

Reply:
> Care to elaborate?

Sure:
a) There is an `exp` leftover that will break the code (Im sure about that)
b)  you reverted `div(fracExp)` with the `mul(siVal.value)` which will lead to different results (we must FIRST multiply with chain decimals and THEN revert the powed frac decimals or else edge scenarios will not work. (Im not 100% sure about that)

>Could you give an example? I don't think I get it.

The symbols we KNOW that are Mul are all capitalcase except 1 (k).
So instead of adding 1 more variable to check against (isMil or isMul) one can check against standard expected values like the check I did before. As soon as the need (at least so far) does not require this extra new boolean to be used across many places, the check against capitalization comparison of self (and k) would be enough I think.

Reply:
> a) There is an exp leftover that will break the code (Im sure about that)

Sure, it was just a suggestion, I obviously didn't check the code just suggested something in Github editor.

> b) you reverted div(fracExp) with the mul(siVal.value) which will lead to different results (we must FIRST multiply with chain decimals and THEN revert the powed frac decimals or else edge scenarios will not work. (Im not 100% sure about that)

We still multiple by decimals and then divide by powered frac, it's ust we don't multiply/divide by `siVal` first because I don't think it's needed (yes, we loose cases where someone writes `0.00<chain-decimals-zeros>1 <largeprefix>`, but as mentioned in other comment it's not something we expected people to write. A limit of `chain-decimals` is a sane one).

> The symbols we KNOW that are Mul are all capitalcase except 1 (k).

This one is about code quality / clarity of intent. Since you've split the place where `siValues` are defined and the place where we have the logic that checks that you create an implicit assumption about that fact. Someone who would add a new unit (let's say `hecto`) will not see that the way we detect mul/div is implicitly based on the assumption about units existing during that particular time. Every kind of unclear assumption like that is not bad per-se, but it's not necessarily future proof - for some cases it makes sense and for some it will bite us in the future. Based on my experience things in the codebase change so frequently that I always try to avoid making any assumptions of this kind and if there are any I write them down in the form of tests or at least documentation.

Reply:
Makes sense. Just when u propose code like that, github gives option of merging, and im always tempted to cllick that merge button. Are you ok with the latest changes? I pretty much followed what u proposed i think